ambry-router/src/main/java/com/github/ambry/router/ParanoidDurabilityOperationTracker.java

```diff
@@ -0,0 +1,405 @@
+    List<ReplicaId> downReplicasToCheck = new ArrayList<>(downReplicas);
+
+    // Add replicas that are neither in local dc nor in originating dc.
+    backupReplicas.forEach(this::addToEndOfPool);
+
+    if(!enoughReplicas()) {
+      // Tommy: Add a simpler exception here that doesn't list out all of the replica information, just refer to Helix.
+      throw new IllegalArgumentException(
+          generateErrorMessage(partitionId, examinedReplicas, replicaPool, backupReplicasToCheck, downReplicasToCheck,
+              routerOperation));
+    }
+
+    initializeTracking();
+  }
+
+  private void initializeTracking() {
+    inflightCountByDc = new HashMap<String, Integer>();
+    replicaSuccessCountByDc = new HashMap<String, Integer>();
+    replicaInPoolOrFlightCountByDc = new HashMap<String, Integer>();
+    remoteDatacenters = new ArrayList<String>();
+
+    replicaPoolByDc.forEach((dcName, v) -> {
+      inflightCountByDc.put(dcName, 0);
+      failedCountByDc.put(dcName, 0);
+      replicaSuccessCountByDc.put(dcName, 0);
+      replicaInPoolOrFlightCountByDc.put(dcName, 0);
+      disabledCountByDc.put(dcName, 0);
+      totalReplicaCountByDc.put(dcName, v.size());
+      allDatacenters.add(dcName);
+      if(!dcName.equals(datacenterName)) {
+        remoteDatacenters.add(dcName);
+      }
+    });
+  }
+
+  // Tommy: From here do we have access to ClusterMapUtils??? NO
+  // How do you construct ClusterMapUtils?
+  private boolean enoughReplicas() {
+    for (String currentDc : replicaPoolByDc.keySet()) {
+      // return false if the local data center has too few replicas (must check that dc name is the local one first)
+      // return false if any SINGLE remote data center has too few replicas
+      if(currentDc.equals(datacenterName)) {
+        if (replicaPoolByDc.get(currentDc).size() < localReplicaSuccessTarget) {
+          logger.error("Not enough replicas in local data center for partition " + partitionId + " to satisfy paranoid durability requirements");
+          return false;
+        }
+      } else {
+        if (replicaPoolByDc.get(currentDc).size() < remoteReplicaSuccessTarget) {
+          logger.error("Not enough replicas in remote data center " + currentDc + " for partition " + partitionId + " to satisfy paranoid durability requirements");
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  public boolean hasSucceeded() {
+    boolean hasSucceeded;
+    // Tommy: The original method in SimpleOperationTracker checks whether or not the current routerOperation is
+    //        a RouterOperation.PutOperation, but we already know that must be true if this class is being used at all.
+    if (routerConfig.routerPutUseDynamicSuccessTarget) {
+      // this logic only applies to replicas where the quorum can change during replica movement
+      int dynamicSuccessTarget = Math.max(totalReplicaCount - disabledCount - 1, routerConfig.routerPutSuccessTarget);
```

Comment:
this is not right, the dynamicSuccessTarget is based on the replica pool size, which is not valid in the `ParanodDuabilityOperationTracker`. What should be compared here is whether we succeed in local and remote colo.

Can we have two methods, `hasSucceededInLocal` and `hasSucceedInRemote`. 

`hasSucceededInRemote` is simple, we just have to compare the remote success target with the remote success count.
`hasSucceededInLocal`, we just have to compare with the local success target that you initialized in the constructor.